src/Pos/Launcher/Runner.hs

```diff
@@ -176,11 +181,34 @@ runRawRealMode peerId transport np@NodeParams {..} sscnp listeners outSpecs (Act
                    runBListenerStub $
                    act
 
-        let startMonitoring node' = case lpEkgPort of
-                Nothing   -> return Nothing
-                Just port -> Just <$> startMonitor port runIO node'
-
-        let stopMonitoring it = whenJust it stopMonitor
+        let startMonitoring node' = case npEnableMetrics of
+                False -> return Nothing
+                True -> Just <$> do
+                    ekgStore <- liftIO $ Metrics.newStore
+                    ekgStore' <- setupMonitor runIO node' ekgStore
+                    liftIO $ Metrics.registerGcMetrics ekgStore'
+                    mEkgServer <- case npEkgParams of
+                        Nothing -> return Nothing
+                        Just (EkgParams {..}) -> Just <$> do
+                            liftIO $ Monitoring.forkServerWith ekgStore' ekgHost ekgPort
+                    mStatsdServer <- case npStatsdParams of
+                        Nothing -> return Nothing
+                        Just (StatsdParams {..}) -> Just <$> do
+                            let statsdOptions = Monitoring.defaultStatsdOptions
+                                    { Monitoring.host = statsdHost
+                                    , Monitoring.port = statsdPort
+                                    , Monitoring.flushInterval = statsdInterval
+                                    , Monitoring.debug = statsdDebug
+                                    , Monitoring.prefix = statsdPrefix
+                                    , Monitoring.suffix = statsdSuffix
+                                    }
+                            liftIO $ Monitoring.forkStatsd statsdOptions ekgStore'
+                    return (mEkgServer, mStatsdServer)
+
+        let stopMonitoring Nothing = return ()
+            stopMonitoring (Just (mEkg, mStatsd)) = do
+                maybe (pure ()) (killThread . Monitoring.statsdThreadId) mStatsd
+                maybe (pure ()) stopMonitor mEkg
```

Comment:
This section is getting rather long. Can we move it out of line, or at least into a where clause?